server/handlers/api_handlers.go

```diff
@@ -0,0 +1,143 @@
+		UUID:     uuid.NewV4(),
+		UserUUID: u.UUID,
+		Secret:   sec.GetRandString(10),
+	})
+
+	if err != nil {
+		logrus.WithError(err).WithField("user_uuid", u.UUID).Error("Error upserting device secret")
+	}
+
+	sendOK(w)
+}
+
+func Login(w http.ResponseWriter, r *http.Request, ps httprouter.Params, d device.Handler, db db.DatabaseHandler, a auth.Authenticator) {
+	req := entities.LoginRequest{}
+	err := json.NewDecoder(r.Body).Decode(&req)
+	// TODO: req validation
+	if err != nil {
+		logrus.WithError(err).Error("Could not parse LoginRequest")
+		err := sendJSON(w, entities.ErrorResponse{Error: "Could not parse LoginRequest"}, 400)
+		if err != nil {
+			logrus.WithError(err).Error("!!!! Could not send error JSON response (LoginRequest)")
+		}
+		return
+	}
+
+	// Get user if exists
+	user, err := db.GetUser(auth.User{Email: req.Email})
+	if err != nil {
+		logrus.WithError(err).Error("Trouble fetching user")
+		err := sendJSON(w, entities.ErrorResponse{Error: "Trouble fetching user"}, 400)
+		if err != nil {
+			logrus.WithError(err).Error("!!!! Could not send error JSON response (Login)")
+		}
+		return
+	}
+
+	// Get user's DeviceSecret to embed into Token
+	secret, err := db.GetDeviceSecret(user.UUID)
+	if err != nil {
+		logrus.WithError(err).WithField("user_uuid", user.UUID).Error("Issue fetching device secret")
+	}
+
+	// Get Token for user
+	token, err := a.GetToken(req.Email, req.Password, &auth.GetTokenOpts{
+		RequestedPermissions: auth.PERMISSIONS_USER,
+		Data:                 auth.TokenData{"deviceSecret": secret.Secret},
+	})
+
+	if err != nil {
+		logrus.WithError(err).Error("Error getting token for user")
+	}
+
+	res := entities.LoginResponse{Token: token}
+	sendJSON(w, res, 200)
+}
+
+func Call(w http.ResponseWriter, r *http.Request, ps httprouter.Params, d device.Handler, db db.DatabaseHandler, a auth.Authenticator) {
+	req := entities.CallRequest{}
+	err := json.NewDecoder(r.Body).Decode(&req)
+	if err != nil {
+		logrus.WithError(err).Error("Could not parse CallRequest")
+		err := sendJSON(w, entities.ErrorResponse{Error: "Could not parse CallRequest"}, 400)
+		if err != nil {
+			logrus.WithError(err).Error("!!!! Could not send error JSON response (CallRequest)")
+		}
+		return
+	}
+
+	// Authenticate User
+	claims, err := a.Validate(req.Token)
+	if err == auth.ErrorValidatingToken {
+		logrus.WithField("token", req.Token).Info("Error validating token")
+		err := sendJSON(w, entities.ErrorResponse{Error: "Error validating token"}, 401)
+		if err != nil {
+			logrus.WithError(err).Error("!!!! Could not send error JSON response (CallRequest)")
+		}
+		return
+	}
+	if err != nil {
+		logrus.WithError(err).Error("Unknown error validating token")
+		err := sendJSON(w, entities.ErrorResponse{Error: "Error validating token"}, 500)
+		if err != nil {
+			logrus.WithError(err).Error("!!!! Could not send error JSON response (CallRequest)")
+		}
+		return
+	}
+
+	d.Call(req.DeviceName, claims.Data["deviceSecret"], req.FunctionName)
```

Comment:
#59

Reply:
Sounds good for v2.1! v2 will be focused on feature parity with v1!